### DISCUSSION TITLE
Pathway to adding missing elements? (e.g. Mg, Si)

### DISCUSSION THREAD

**Discussion by orionarcher** [openforcefield/discussions#26] [Q&A] [answered]
I am an avid user of the OpenFF tools but coming from the battery space, many elements I am interested in are missing from the Sage FF. To name a few, Mg (for Mg-ion batteries) and Si (common in additives). I'm curious what a viable path to integrating these with Sage would be. I am considering:

1. Using parameters from GAFF, possibly paired with BespokeFit for Si dihedrals. This would require converting the GAFF parameters to SMARTs. Would y'all expect this to work? I personally have more faith in the Mg parameters than the Si parameters but I'd love thoughts from someone more experienced.

2. Trying to train Mg or Si parameters from scratch. I'm not sure what this would entail and would be curious what official tooling exists for integrating new elements into the FF.

3. Using GAFF in it's entirety. Eliminates the need for FF mixing but GAFF is a bit more antiquated imho.

**Comment by mattwthompson** [comment, marked as answer]
For background and for anybody who is just coming across this topic, (@orionarcher I figure you're already aware of the this) the official stance on adding these elements into the mainline OpenFF force fields is currently something like Mobley's comment [here](https://github.com/orgs/openforcefield/discussions/19). I'm not sure we have a policy in place for external contributions like if you were to train Mg and Si parameters; this sounds like one of those "good problems" one wishes to have one day.

For anybody else who wants to build off of a mainline OpenFF force field (Parsley, Sage, sometime later to include Rosemary and Thyme) and make their own SMIRNOFF force field using OpenFF tools, I think there are generally two categories of solutions here:

1. The "easy" route, which doesn't require any changes to the functional form, etc., "just" more data. This would entail some amount of generating QM data and/or curating physical property data, and if the data is there, it could be as simple as running the fits again with extra parameters and extra training data. There are some number of elements (boron comes to mind?) that fit into this category right now, either due to insufficient data or relevance to the current applications.
3. The "hard" route, which involves changes to functional forms or the infrastructure. I think, for example, there's some doubt as to how well point charges can handle things like large trivalent ions in water or metalloproteins. Cases like these would be a heavier lift, though some intrepid explorers have done this via [`SMIRNOFF plugins`](https://github.com/openforcefield/smirnoff-plugins).

Back to your question:

* Porting an entire Amber-style general force field to SMIRNOFF is considered impossible (a similar attempt was made to [port GAFF to Foyer](https://github.com/rsdefever/GAFF-foyer), which you'd think is possible, but was never able to get to 100% due to some mysterious details). However, it might be possible for some limited chemistries, like if you're only looking at one specific molecule and you already know what parameters should end up where. In that case it can "work" in the sense that the same parameters will be applied to the molecule you wrote the SMARTS patterns for, but this breaks down somewhere between a single molecule and all of chemical space, so it's not an advisable path to go down if you're interested in something like high-throughput screening.
* For aqueous magnesium, are there any published parameter sets that are fit with TIP3P? The ion parameters in Sage are not refit, they are from Juong and Chetham 2008. If you can find something, you should be able to just add it in and get results as good as those parameters imply.
  * Chapin recently added a SMIRNOFF version of OPC to the `openff-forcefields` repo. I'm not suggesting you use this for your work (OPC is a 4-site water model and Sage is not trained to be compatible with it), but this is an example of how straightforward it is to use existing parameters in our infrastructure.
  * Even if you wanted to generate new parameters, I think it'd be easier to grab some commonly-used values and retrain them with new data vs. create new ones from scratch.
* Silicon is somewhat [esoteric](https://www.science.org/content/blog-post/silicon-stays-shadows) in the world of drug discovery. In principle if you're using it in things that otherwise look like normal molecules (i.e. not a salt, solid electrolyte, or something eager to react) if a carbon was in its place and if you really only need to fit one torsion, it'd be worth trying BespokeFit. My concern here is that the non-bonded parameters are poorly compatible with Sage and that refitting the torsion to match the QM profile isn't enough to get the rest of the molecule to behave well. I have no idea if it would work, but it'd be interested to see what happens if you try!
  * Refitting a _general_ force field with silicon would require a lot of data that might be harder to get, at least I'm guessing.

**Reply by orionarcher** [reply]
Thanks for the quick and thorough answer @mattwthompson. 

> If you can find something, you should be able to just add it in and get results as good as those parameters imply.

I've found several recent citations for Mg parameters fit with TIP3P. It's very helpful to know those are effectively a drop in solution. Where could I find examples/documentation for adding literature parameters into the existing Sage FF?

>There are some number of elements (boron comes to mind?) that fit into this category right now, either due to insufficient data or relevance to the current applications.

I'll definitely consider the "easy" option here. I'll need to read up on data set generation for Sage but a QM campaign may be tractable. Boron is of interest for me as well.

**Reply by mattwthompson** [reply]
Historically I've done this by hand, often by copying existing lines in the XML and updating them. If you want to do things in a more systematic way in memory, the API for adding parameters is `ParameterHandler.add_parameter` ([example](https://github.com/openforcefield/openff-toolkit/blob/stable/examples/forcefield_modification/forcefield_modification.ipynb)).

Note that you can load multiple force field files at once, i.e.

```python
ForceField("openff-2.0.0.offxml", "tip4p.offxml")
```
or
```python
ForceField("openff-2.0.0.offxml", "orions_silicon_parameters.offxml")
```

so in the case of Mg2+ parameters, you could even have them in a file that contains nothing but its vdW and charge parameters (~10 lines total with all of the metadata).

**Comment by davidlmobley** [comment]
As a note, we're also very interested in Si parameters in general (and all of the monovalent ions), but haven't had time/bandwidth to task someone with developing these specifically for Sage. For Si, a pathway to getting this into future force fields would be (a) developing starting point parameters via a route like you suggest (with BespokeFit) and then handing them over to us, which would solve our current chicken-and-egg problem of "we need starting point parameters before we can routinely refit them as part of our upcoming releases."  (There is also a question, for Si, of where the LJ parameters would come from -- what condensed phase properties we would use.)

Ions are a little harder, potentially.

Happy to discuss if this is something you want to push forward on, but otherwise the approach that Matt discussed will work for your personal use.
